zeusd/src/devices/cpu/mod.rs

```diff
@@ -0,0 +1,170 @@
+use std::path::PathBuf;
+use std::sync::RwLock;
+use std::time::Instant;
+use tokio::sync::mpsc::{Sender, UnboundedReceiver, UnboundedSender};
+use tracing::Span;
+
+use crate::error::ZeusdError;
+
+pub struct PackageInfo {
+    pub index: u32,
+    pub name: String,
+    pub energy_uj_path: PathBuf,
+    pub max_energy_uj: u64,
+
+    num_wraparounds: RwLock<u64>,
+}
+
+#[derive(Serialize)]
+pub struct RaplResponse {
+    cpu_energy_uj: Option<u64>,
+    dram_energy_uj: Option<u64>,
+}
+
+pub trait CpuManager {
+    fn device_count() -> Result<u32, ZeusdError>;
+    fn get_available_fields(index: u32) -> Result<(PackageInfo, Option<PackageInfo>), ZeusdError>;
+    fn get_cpu_energy(&self) -> Result<u64, ZeusdError>;
+    fn get_dram_energy(&self) -> Result<Option<u64>, ZeusdError>;
+    fn stop_monitoring(&mut self);
+}
+
+pub type CpuCommandRequest = (
+    CpuCommand,
+    Option<Sender<Result<RaplResponse, ZeusdError>>>,
+    Instant,
+    Span,
+);
+
+#[derive(Clone)]
+pub struct CpuManagementTasks {
+    // Senders to the CPU management tasks. index is the CPU ID.
+    senders: Vec<UnboundedSender<CpuCommandRequest>>,
+}
+
+impl CpuManagementTasks {
+    pub fn start<T>(cpus: Vec<T>) -> anyhow::Result<Self>
+    where
+        T: CpuManager + Send + 'static,
+    {
+        let mut senders = Vec::with_capacity(cpus.len());
+        for (cpu_id, cpu) in cpus.into_iter().enumerate() {
+            // Channel to send commands to the CPU management task.
+            let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+            senders.push(tx);
+            // The CPU management task will automatically terminate
+            // when the server terminates and the last sender is dropped.
+            tokio::spawn(cpu_management_task(cpu, rx));
+            tracing::info!("Background task for CPU {} successfully spawned", cpu_id);
+        }
+        Ok(Self { senders })
+    }
+
+    pub async fn send_command_blocking(
+        &self,
+        cpu_id: usize,
+        command: CpuCommand,
+        request_start_time: Instant,
+    ) -> Result<RaplResponse, ZeusdError> {
+        if cpu_id >= self.senders.len() {
+            return Err(ZeusdError::CpuNotFoundError(cpu_id));
+        }
+        let (tx, mut rx) = tokio::sync::mpsc::channel(1);
+        self.senders[cpu_id]
+            .send((command, Some(tx), request_start_time, Span::current()))
+            .unwrap();
+        match rx.recv().await {
+            Some(result) => result,
+            None => Err(ZeusdError::CpuManagementTaskTerminatedError(cpu_id)),
+        }
+    }
+
+    pub async fn stop_monitoring(&self) {
+        for sender in self.senders.clone() {
+            let (tx, _) = tokio::sync::mpsc::channel(1);
+            sender
+                .send((
+                    CpuCommand::StopMonitoring {},
+                    Some(tx),
+                    Instant::now(),
+                    Span::current(),
+                ))
+                .unwrap();
```

Comment:
Since `CpuCommand::execute` will return a bogus `RaplResponse` when monitoring is stopped, maybe `recv` that bogus response and discard it here? That'll ensure that `stop_monitoring` will wait until the termination of wraparound monitors are actually issued before returning.